src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived, or specifies
+     *     a type of output that is not a key (e.g. raw data)
+     * @throws IllegalArgumentException
+     *     if {@code alg} is {@code null} or empty
+     */
+    @Override
+    protected SecretKey engineDeriveKey(String alg,
+                                        AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException {
+
+        if (alg == null || alg.isEmpty()) {
+            throw new IllegalArgumentException(
+                "the algorithm for the resultant SecretKey may not be null or"
+                + " empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(derivationParameterSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived key material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        SecretKey inputKeyMaterial;
+        SecretKey salt;
+        SecretKey pseudoRandomKey;
+        byte[] info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (derivationParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) derivationParameterSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // perform extract
+            try {
+                return hkdfExtract(inputKeyMaterial,
+                                   (salt == null) ? null : salt.getEncoded());
```

Comment:
RFC 5869 sec2.2 states that "salt - if not provided, it is set to a string of HashLen zeros." To match the spec better, we should do this instead of use `null`?